### PR TITLE
Table.group_by

### PR DESCRIPTION
**Pull request by jdunkerley**
### Pull Request Description

Functioning group_by based of Enso Map.

### Important Notes

This is an initial version which will be used to establish the API.
The grouping map will need to be moved to Java code for performance.

### Checklist

Please include the following checklist in your PR:

- [x] The documentation has been updated if necessary.
- [x] All code conforms to the [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md), [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md), and [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guides.
- All code has been tested:
  - [x] Unit tests have been written where possible.
  - [ ] If GUI codebase was changed: Enso GUI was tested when built using BOTH `./run dist` and `./run watch`.

### CODE REVIEW

File: distribution/lib/Standard/Table/0.0.0-dev/src/Data/Aggregate_Column.enso [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+    type Count (name:Text|Nothing=Nothing)
+
+    ## Creates a new column with the count of unique items in the selected
+       column(s) within each group.
+    # type Count_Distinct (columns:Column|Text|Integer|[Column|Text|Integer]) (name:Text|Nothing=Nothing)
+    type Count_Distinct (columns:Column|Text|Integer|Vector) (name:Text|Nothing=Nothing)
+
+    ## Creates a new column with the count of not null (Nothing) values of the
+       specified column within each group.
+    type Count_Not_Null (column:Column|Text|Integer) (name:Text|Nothing=Nothing)
```
Comment (radeusgd) [inline distribution/lib/Standard/Table/0.0.0-dev/src/Data/Aggregate_Column.enso]: Probably should be updated to match the updated design (`Count_Not_Nothing` etc.)

File: distribution/lib/Standard/Table/0.0.0-dev/src/Data/Aggregate_Column.enso [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+
+## Defines an Aggregate Column
+type Aggregate_Column
+    ## Creates a new column with the row count of each group
+    type Count (name:Text|Nothing=Nothing)
+
+    ## Creates a new column with the count of unique items in the selected
+       column(s) within each group.
+    # type Count_Distinct (columns:Column|Text|Integer|[Column|Text|Integer]) (name:Text|Nothing=Nothing)
+    type Count_Distinct (columns:Column|Text|Integer|Vector) (name:Text|Nothing=Nothing)
```
Comment (radeusgd) [inline distribution/lib/Standard/Table/0.0.0-dev/src/Data/Aggregate_Column.enso]:
```suggestion
    type Count_Distinct (columns:Column|Text|Integer|[(Column|Text|Integer)]) (name:Text|Nothing=Nothing)
```

I think adding these additional parentheses could make the parser accept this new syntax.

File: distribution/lib/Standard/Table/0.0.0-dev/src/Data/Aggregate_Column.enso [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+    type Median (column:Column|Text|Integer) (name:Text|Nothing=Nothing)
+
+    ## Creates a new column with the standard deviation of values (ignoring
+       NULLs) of the column within each group.
+
+       Arguments:
+       - population argument specifies if group is a sample or the population
+    type Standard_Deviation (column:Column|Text|Integer) (name:Text|Nothing=Nothing) (population:Boolean=False)
+
+    ## Creates a new column with the values concatenated together. NULL values will become an empty string.
```
Comment (radeusgd) [inline distribution/lib/Standard/Table/0.0.0-dev/src/Data/Aggregate_Column.enso]: Btw. why is NULL collated with `""` here instead of being ignored like in other cases ?

File: distribution/lib/Standard/Table/0.0.0-dev/src/Data/Aggregate_Column.enso [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+            Count_Not_Empty c _ -> create_closure c col->count->i->(count + if is_empty (col.at i) then 0 else 1)
+            Count_Empty c _ -> create_closure c col->count->i->(count + if is_empty (col.at i) then 1 else 0)
+            First c _ -> create_closure c col->current->i->(if i==0 then (col.at i) else current)
+            Last c _ -> create_closure c col->_->i->(col.at i)
+            Sum c _ -> create_closure c col->total->i->
+                v = col.at i
+                (total + if v.is_nothing then 0 else v)
+            Average c _ -> create_closure c col->a->i->
+                v = col.at i
+                (c + if v.is_nothing then a else [(a.at 0) + 1, (a.at 1) + v])
```
Comment (radeusgd) [inline distribution/lib/Standard/Table/0.0.0-dev/src/Data/Aggregate_Column.enso]:
```suggestion
                (c + if v.is_nothing then a else [a.first + 1, a.second + v])
```

I think it may be worth using these two helper functions, will be clearer to read (at least in my opinion - because when I see `at` I usually think it may be an arbitrary size array and if I see `first` and `second` I immediately connect this with the accumulator which is a pair).

File: distribution/lib/Standard/Table/0.0.0-dev/src/Data/Aggregate_Column.enso [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+                if i==0 then val else (text + j + val)
+            Median c _ -> create_closure c col->map->i->
+                val = col.at i
+                if val.is_nothing then map else (map.insert val (1 + (map.get_or_else val 0)))
+            Count_Distinct columns _ ->
+                resolved = columns.map c->(this.resolve_column table c)
+                key_maker i = Group_By_Key (resolved.map c->(c.at i))
+                map->i->(map.insert (key_maker i) 1)
+
+    evaluate : Any->Any
```
Comment (radeusgd) [inline distribution/lib/Standard/Table/0.0.0-dev/src/Data/Aggregate_Column.enso:181]: I'm a bit skeptical about this split of `initial_value`, `aggregator` and `evaluate` - requires a lot of 'looking around' to understand the algorithm for a given function.

As this is a prototype and will be changed in the future - I do _not_ suggest to change anything now.

But in the future for similar use cases, I would suggest doing this in a more "local" way, for example we can define a helper type:
```
type Aggregation_Algorithm initializer aggregator evaluator
```

and then create the algorithms in one place:
```
algorithm : Aggregation_Algorithm
algorithm = case this of
    ...
    Last _ -> 
        initializer = ...
        aggregator = ...
        evaluator = ...
```

This way all parts of each algorithm are next to each other so it is much easier to grasp a given algorithm and see if it is correct. If the algorithm is spread throughout multiple places it is harder to spot potential issues.

File: distribution/lib/Standard/Table/0.0.0-dev/src/Data/Aggregate_Column.enso [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+        Median _ _ -> (Error.throw "Not in prototype")
+        Average _ _ -> if (value.at 0) == 0 then Nothing else (value.at 1 / value.at 0)
+        Standard_Deviation _ _ p -> if (value.at 0) == 0 then Nothing else
+            (value.at 2/value.at 0 - (value.at 1/value.at 0)^2).sqrt * (if p then 1 else 1)
+        Concatenate _ _ _ s p -> if value.is_nothing then value else (s + value + p)
+        _ -> value
+
+
+## Occurs when cannot aggregate a column
+type Invalid_Aggregation_Method (column : Text) (message : Text)
```
Comment (radeusgd) [inline distribution/lib/Standard/Table/0.0.0-dev/src/Data/Aggregate_Column.enso:211]: I would suggest rewording this as it seems to be missing a subject? Maybe just make it passive voice.

File: distribution/lib/Standard/Table/0.0.0-dev/src/Data/Group_By.enso [hunk 7]
```diff
@@ -0,0 +1,25 @@
+from Standard.Base import all
+
+## Group By Key
+type Group_By_Key
```
Comment (radeusgd) [inline distribution/lib/Standard/Table/0.0.0-dev/src/Data/Group_By.enso]: This comment just repeats the name. IMO we shouldn't have comments like this, because they do not really add anything. I'm not really sure what to write here, but it just looks weird. Maybe in this case we can just have no comment at all... But maybe we can write some intention here, for example 'An internal key used for performing the group by operation in memory.' or something like that.

File: distribution/lib/Standard/Table/0.0.0-dev/src/Data/Group_By.enso [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+    type Group_By_Key values
+
+    ## See if two keys are equal
+    == : Group_By_Key->Boolean
+    == that =
+        if that.values.length != this.values.length then False else
+            0.up_to this.values.length . all i->(this.values.unsafe_at i == that.values.unsafe_at i)
```
Comment (radeusgd) [inline distribution/lib/Standard/Table/0.0.0-dev/src/Data/Group_By.enso]: `unsafe_at` suggests that `values : Vector`.
But this implementation of `==` is exactly the same as `Vector.==`. So why override it?
I think we can just rely on the autogenerated one, as it will have the same semantics, or if you really want to explicitly write it, at least do:
```suggestion
    == : Group_By_Key->Boolean
    == that =
        that.values == this.values
```

Unless there is some reason that I'm not seeing as to why it should be implemented separately?

File: distribution/lib/Standard/Table/0.0.0-dev/src/Data/Group_By.enso [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+
+    ## See if two keys are not equal
+    != : Group_By_Key->Boolean
+    != that = this==that . not
```
Comment (radeusgd) [inline distribution/lib/Standard/Table/0.0.0-dev/src/Data/Group_By.enso]: Again, this is already automatically handled by the definition at `Any.!=`. No need to repeat it (or were there issues without this?? Then I think it would be a bug worth reporting)

File: distribution/lib/Standard/Table/0.0.0-dev/src/Data/Group_By.enso [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+
+    ## Compares two keys
+    compare_to : Group_By_Key->Ordering
+    compare_to that =
+        loop i =
+            comparison = this.values.unsafe_at i . compare_to (that.values.unsafe_at i)
+            if comparison != Ordering.Equal then comparison else
+                if i == this.values.length - 1 then Ordering.Equal else
+                    @Tail_Call loop (i+1)
+        loop 0
```
Comment (radeusgd) [inline distribution/lib/Standard/Table/0.0.0-dev/src/Data/Group_By.enso]: Again, cannot we just reuse `Vector_Lexicographic_Order.compare`? Seems like this is the same algorithm modulo the fact that this algorithm does not check if both vectors have the same length so it can crash in a nasty way (as it uses `unsafe_at`) if they do not. I know normally we assume the group keys will be of the same length, but IMO if using `unsafe_at` the check should be there (and interestingly, it is there for `==`).

File: distribution/lib/Standard/Table/0.0.0-dev/src/Data/Table.enso [hunk 11]
```diff
@@ -507,6 +510,52 @@ type Table
...
+            key = make_key i
+            row_index = map.get_or_else key (add_row key)
+
+            # Accumulate...
+            0.up_to (columns.length) . each j->
+                aggregator = aggregators.at j
+                array = new_table.at (j + key_length) . at 1 . to_array
+                current = array . at row_index
+                new = aggregator current i
+                array . set_at row_index new
```
Comment (radeusgd) [inline distribution/lib/Standard/Table/0.0.0-dev/src/Data/Table.enso:546]: This may stop working when we fix the "`to_array` is leaking mutability" issue.

As this prototype may live shorter than the above issue - probably ignore this comment.

But just for future - maybe we could just create an `Expandable_Array` which wraps a regular, explicitly-mutable array but also has the `append` operation and does the allocations that give us O(1) amortized complexity of `append`. Because I don't like that we are abusing the `Vector_Builder` a bit when we could just create a data structure meant for such use-cases. But since this is a prototype - just commenting a suggestion for future.

**Review by radeusgd — APPROVED** [review]
Looks good to me.
Some style comments and questions but given it's a prototype, most are just suggestions for the future.

The important thing is the `Group_By_Key` - IMO it should not repeat the implementations of existing functions if it can delegate to them _unless_ it is really absolutely necessary.

File: distribution/lib/Standard/Table/0.0.0-dev/src/Data/Aggregate_Column.enso (hunk 5 above)
Comment (jdunkerley) [reply distribution/lib/Standard/Table/0.0.0-dev/src/Data/Aggregate_Column.enso:181]: Added as a follow on ticket.

File: distribution/lib/Standard/Table/0.0.0-dev/src/Data/Aggregate_Column.enso [hunk 12]
```diff
@@ -77,19 +77,22 @@ type Aggregate_Column
...
           results within the group.
-    type Last (column:Column|Text|Integer) (name:Text|Nothing=Nothing) (ignore_nothing:Boolean=False) (order_by:Column_Selector|Nothing=Nothing)
+    type Last (column:Column|Text|Integer) (name:Text|Nothing=Nothing) (ignore_nothing:Boolean=True) (order_by:Column_Selector|Nothing=Nothing)
 
     ## Gets a column name to use for the aggregate column
     column_name : Table->Text
     column_name table =
         if this.name.is_nothing.not then this.name else
-            get_name c = (here.resolve_column table c).name
+            get_name c = (this.resolve_column table c).java_column.getName
```
Comment (radeusgd) [inline distribution/lib/Standard/Table/0.0.0-dev/src/Data/Aggregate_Column.enso]: Why not use `.name`??

The naming logic should be shared between backends, but this method will not work with DB columns which do not have a `java_column`. But they do have a `name`.

File: distribution/lib/Standard/Table/0.0.0-dev/src/Data/Group_By.enso [hunk 13]
```diff
@@ -1,19 +1,54 @@
...
+
+## PRIVATE
+   Represents a comparable vector of element which is used as key for grouping.
 type Group_By_Key
     type Group_By_Key values
 
     ## See if two keys are equal
     == : Group_By_Key->Boolean
-    == that = this.values == that.values
+    == that = (this.compare_to that) == Ordering.Equal
```
Comment (radeusgd) [inline distribution/lib/Standard/Table/0.0.0-dev/src/Data/Group_By.enso]: Why not just `==`? I think the two should have the same semantics.

**Review by radeusgd — APPROVED** [review]
Looks good to me
